src/hotspot/share/utilities/events.cpp

```diff
@@ -47,19 +46,22 @@ StringEventLog* Events::_class_loading = nullptr;
 StringEventLog* Events::_deopt_messages = nullptr;
 StringEventLog* Events::_dll_messages = nullptr;
 
+static EventLog* _event_logs = nullptr;
+
 EventLog::EventLog() {
-  // This normally done during bootstrap when we're only single
-  // threaded but use a ThreadCritical to ensure inclusion in case
-  // some are created slightly late.
-  ThreadCritical tc;
-  _next = Events::_logs;
-  Events::_logs = this;
+  // This normally done during bootstrap when we're only single threaded,
+  // but use lock free add because there are some events that are created later.
+  EventLog* old_head;
+  do {
+    old_head = Atomic::load(&Events::_logs);
+    _next = old_head;
+  } while (Atomic::cmpxchg(&Events::_logs, old_head, this, memory_order_relaxed) != old_head);
```

Comment:
I don't think it is safe to use `memory_order_relaxed` order if events are added after JVM enters multi-thread mode.

Reply:
Thanks for this. I didn't know why Aleksey suggested it.  I'll remove it.

Reply:
Maybe @shipilev meant `memory_order_release`? Anyway, I guess we don't need to optimize it.

Reply:
I saw no point in enforcing memory ordering mode here, as it looks like we only did `ThreadCritical` for mutual exclusion. Note that we do not have a matching acquire on list traversals, so we seqcst/release on list additions would be incomplete. That only reinforces my original thinking: we are riding on memory ordering given by something else, I'd guess the initialization sequence itself.

But I won't quibble, it is a very minor optimization.

Reply:
I guess we rely on memory ordering by address dependency on the reader's side.

Reply:
Yes we only used ThreadCritical to add to the list after the VM becomes multithreaded.  The list traversals (reads) are for hs_err_pid file printing, which presumably is single threaded at that point and presumably nothing is adding to the list.  I added Atomic::load() but I think it's not necessary to be a load_acquire.
There are only about 10 items max on this list so far, so I think performance isn't a concern either way.

Reply:
There is also a jcmd to print out events. In theory, that could be a problem if  CAS is unordered.

Reply:
hm then should the readers be Atomic::load_acquire() ?

Reply:
The version you merged has `memory_order_conservative` order on CAS, so you don't need reader side barrier.

Reply:
`memory_order_conservative` is [Strong two-way memory barrier](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/atomic.hpp#L47), that means all reads and writes *cannot* float pass the barrier.

If `Atomic::load()` observes a store by conservative order, then precedent reads and writes of the store must have completed.

Reply:
First of all, nothing on the writer's side can fix the reader's side.

We're basically relying on volatile load + dependency chain on the reader's side. This is very likely to work, but C++ memory model specialists may find this questionable.

Reply:
Are we on the same page? which `load` you are talking about?

Reply:
I'm talking about `Atomic::load()` which is essentially a volatile load without ordering.

Reply:
I don't believe we need reader side barrier, given the `conservative` order on writer side. @shipilev any opinion?

Reply:
Well, if we are going fully pedantic here, we only need to make sure that the `EventLog::_next` is properly visible to the code that traverses the `EventLog` list. This realistically only matters if there are no synchronization points in between those, which I suspect are plenty since init to any reasonable events log use. I am pretty sure this is the same thing that make all other init code robust in practice: things that happen at init are good to go.

If you want to cover the path to be fully 100% bullet-proof and squeeze some performance at the same time, then I think `cmpxchg` to `Events::_logs` should be `release` and loads from there should be `acquire`-s.